Example of a chaincode using role-based access control (RBAC)

adecaro opened the pull request:
The PR contains a chaincode showing how to implement role-based access control using TCerts.
This addresses Issue #1207.

Signed-off-by: Angelo De Caro adc@zurich.ibm.com

adecaro committed:
Chaincode showing how to implement role-based access control using TCerts

srderson commented:
Why does rbac.yaml contain a complete copy of core.yaml? My concern is that we will not be able to keep these in sync.

Also, is it possible to include a README with this chaincode describing its functionality? I'd preferably like to do that for all example chaincodes.

adecaro commented:
Hi @srderson, this chaincode is part of a joint effort with @jeffgarratt to write behave tests for the access control mechanism supported by the fabric.

A copy of core.yaml is there for test purposes. It will not stay there forever. In fact, once the client sdk will be ready I will write an app to show how to invoke the chaincode leveraging the client sdk drectly. Actually, having a test that shows how to invoke such a chaincode will help us driving the design of the client sdk to ensure that all the required API are exposed as expected. What do you think?

For the README, I will write it soon :)

srderson commented:
@adecaro @jeffgarratt So it looks like this chaincode is deployed in the `rbac_test.go` test case? Is it possible to use a behave test instead? The unit tests where chaincodes are deployed are generally fragile. Ideally I think they should all be moved over to behave tests.

adecaro commented:
@srderson, Indeed, rbac_test simulates a minimal fabric environment, deploys the chaincode and invokes it in order to test its behavior. Actually, the final goal is to have behave tests. I'm just concerned that without the client sdk, it would be difficult to run successfully any behave test because the CLI (nor the REST API) does not provide the required API to drive the chaincode.

Anyway, I will try to understand if we can have something cleaner without waiting the client sdk before merging :)

jeffgarratt commented:
@adecaro @srderson The behave test could be implemented as part of the effort I am working on now.  Either #1139 or #961.

binhn commented:
@srderson  should we wait for the behave tests to merge or just merge it since it is a example chaincode?

srderson commented:
@binhn I'd like to avoid merging `rbac.yaml` and `rbac_test.go`. The current effort is to remove these types of tests, not add more. Also, I'd really like to see a README with all example chaincodes, at least documenting their purpose and API.

adecaro commented:
@binhn @srderson  I also think we can wait the behave tests to merge. I'm working with @jeffgarratt to have them ready asap.

christo4ferris commented:
@adecaro could you please fix the license in these new files? It should be the text that I have in PR #1288 thanks

jeffgarratt commented:
@adecaro @srderson I need to touch base with @adecaro and work this workflow.  Have come to a block point based upon current flow (which is already a bit precarious).

adecaro committed:
Updating license.

adecaro committed:
Removing test cases. They will be replaced by behave tests.

adecaro committed:
Adding readme file.

adecaro commented:
Hi @christo4ferris, I have updated the license and also added a readme file to describe the chaincode. 

@christo4ferris, @jeffgarratt, @srderson I can also add a demo app like I did for the asset management chaincode to show how to invoke it. What do you think?

BiscuitBomber commented on examples/chaincode/go/rbac/README.md:
```diff
@@ -0,0 +1,82 @@
+1. Alice is the deployer of the chaincode;
+2. Alice wants to assign the administrator role to Bob;
+3. Alice obtains, via an out-of-band channel, a TCert of Bob, let us call this certificate *BobCert*;
+4. Alice constructs a deploy transaction, as described in *application-ACL.md*,  setting the transaction metadata to *DER(BobCert)*.
```
Please can we make this a hyperlink/url too ?

christo4ferris replied:
could you please clarify... you want a link to this README from where? (I am assuming that is what you're asking for here)

BiscuitBomber replied:
Sorry the suggestion wasn't clear. I meant to suggest: "as described in [Application ACLs](https://github.com/hyperledger/fabric/blob/master/docs/tech/application-ACL.md)"
 instead of "as described in ***application-ACL.md***" [in line 33]

That is:  [Application ACLs] (https://github.com/hyperledger/fabric/blob/master/docs/tech/application-ACL.md)

christo4ferris replied:
got it thanks for the clarification - keep up the great feedback!

adecaro replied:
@BiscuitBomber @christo4ferris Fixed. Thanks for the feedback :)

BiscuitBomber commented on examples/chaincode/go/rbac/rbac.go:
```diff
@@ -0,0 +1,305 @@
+		return nil, errors.New("Failed getting metadata.")
+	}
+	if len(deployer) == 0 {
+		return nil, errors.New("Invalid admin certificate. Empty.")
```
"Empty.": Maybe "Missing caller metadata in admin certificate", "Deployer information is required in admin certificate", or so?

(trying to be a bit more informative in reporting the error.)

adecaro replied:
Hi @BiscuitBomber, what do you think of the following: "Missing metadata. Admin certificate expected in the metadata field."?

BiscuitBomber replied:
Oh, I see - I have learned something new from your proposed change already.

How about "An admin certificate must be provided in the metadata field."? Although I don't have a strong preference either way, so long as the message is instructive.

adecaro replied:
Deal :)

adecaro committed:
Fixing hyperlinks and chaincodeInput typo.

adecaro committed:
Changing error message when an admin certificate is not provided in the metadata field at initialization time.

jeffgarratt commented:
@adecaro @srderson @binhn Want to hold this PR until it is reconciled with TCert with attrs example.

ghaskins commented:
Can this be closed and then reopened when its ready?

christo4ferris commented:
this needs some unit/function tests

ghost commented:
Can one of the admins verify this patch?
